### ISSUE TITLE
AxeBuilder.WithTags not analyzing rules of tags 'wcag2a' and 'wcag21aa'

### ISSUE THREAD

**Issue by pawlaguna** [TroyWalshProf/SeleniumAxeDotnet#9]
Hey all, my objective is to scan all the pages for our company site. Our target compliance level is A. So, naturally I'd like to only scan for level A issues. When I use this line of code...

`var results = new AxeBuilder(driver).WithTags("wcag2aa").Analyze();`

I'll get all wcag2aa level complaints. Great! That's working by design.
Output...
`Running Axe.Axe.Axe.SearchFunctionality.TCID2ScanPublicPage ...`
`Expected results.Violations to be empty, but found {Selenium.Axe.AxeResultItem`
`   {`
     ` Description = "Ensures the contrast between foreground and background colors meets WCAG` `2 AA contrast ratio thresholds"`
      `Help = "Elements must have sufficient color contrast"`
     ` HelpUrl = "https://dequeuniversity.com/rules/axe/3.3/color-contrast?application=axeAPI"`
     ` Id = "color-contrast"`
     ` Impact = "serious"`
     ` Nodes = {Selenium.Axe.AxeResultNode`
With the entire laundry list of color contrast issues.

 However, when I set the tag to `wcag2a` 
`var results = new AxeBuilder(driver).WithTags("wcag2a").Analyze();`

I get back... 
Output...
`Running Axe.Axe.Axe.SearchFunctionality.TCID2ScanPublicPage ...`
`Selenium.Axe.AxeResult`
`No test is available in /Users/plaguna/Projects/Axe/AutomationResources/bin/Debug/netcoreapp2.2/AutomationResources.dll. Make sure that test discoverer & executors are registered and platform & framework version settings are appropriate and try again.`

When I change the tags to `wcag21aa` and or `section508`, I get the same test result output view from the console.

Am I using the tags incorrectly, or could this be a legitimate issue?
I can send more information if need :)

Thanks!

`VS Community/v8.2.5/OSX`
`Selenium 3.14`
`Microsoft.NETCore.App 2.2.0`
`FluentAssertions 5.9.0`
`Microsoft.NET.Test.Sdk 16.0.1`
`MSTest.TestAdapter 1.4.0`
`MSTest.TestFramework 1.4.0`
`Selenium.Axe 1.2`
`Selenium.Support 3.14.1`
`Selenium.WebDriver.ChromeDriver 76.0.3809.12600`

**Comment by pawlaguna** [comment]
That makes sense @dbjorge.
Thank you for the prompt reply.
There's two way I ran the test
1) From the VS IDE I ran the test by "Unit Test" -> "Run All" and received the results in question.
2) And from the terminal, I ran `dotnet restore` -> `dotnet build` -> `dotnet test`

I've uploaded the full test file and .csproj here in GitHub. https://github.com/pawlaguna/MSWPScan

Thanks again!

**Comment by dbjorge** [comment]
* You can suppress the `No test is available...` message by adding `<IsTestProject>false</IsTestProject>` as a child of the first `<PropertyGroup>` in `AutomationResources.csproj` (and any other project that does not contain test cases)
* The color contrast issues not showing up when testing against `wcag2a` is by design because [the motivating WCAG rule (1.4.3)](https://www.w3.org/TR/UNDERSTANDING-WCAG20/visual-audio-contrast-contrast.html) is a AA criteria, not an A criteria. If you'd like more information about which rules WCAG considers to be A vs AA, this [quickref](https://www.w3.org/WAI/WCAG21/quickref) page they provide is a good resource to check.
* Your `AxeScanLevelAA` method is a little off; it's including only the tag `"wcag21aa"`, but this tag only contains rules applicable to WCAG rules that are new between WCAG 2.1 vs 2.0; you probably want that method to include `WithTags("wcag2aa", "wcag21aa")` to get all current AA applicable rules
* The extraneous `Selenium.Axe.AxeResult` output line you're seeing is because of the `Console.WriteLine`s of the result object; the result object doesn't currently provide a custom `ToString`, so these lines aren't going to be useful.
* I checked out the repo you linked, modified the test to scan for "wcag2a" issues rather than the "wcag21aa" it was originally using, and it looks like the test is finding the expected set of `object-alt` violations on the page under test (which is a level A rule)

I'm not sure this has answered your question, but it looks like everything is working by design to me. Can you help me understand more concretely exactly what you believe the problem to be?

**Comment by pawlaguna** [comment]
Thank you for the detailed information here.

- You can suppress the No test is available... message by adding <IsTestProject>false</IsTestProject> as a child of the first <PropertyGroup> in AutomationResources.csproj (and any other project that does not contain test cases)

That's very helpful thank you. I'll use that moving forward.

- The color contrast issues not showing up when testing against wcag2a is by design because the motivating WCAG rule (1.4.3) is a AA criteria, not an A criteria. If you'd like more information about which rules WCAG considers to be A vs AA, this quickref page they provide is a good resource to check.

Yes, that is correct and I'm aware that color-contrast issues are only part of wcagAA and AAA. But I do appreciate the extra context here and the quickref page you've added. Apologizes for the confusion on that part. This is definitely great information for me to run through again though.

- Your AxeScanLevelAA method is a little off; it's including only the tag "wcag21aa", but this tag only contains rules applicable to WCAG rules that are new between WCAG 2.1 vs 2.0; you probably want that method to include WithTags("wcag2aa", "wcag21aa") to get all current AA applicable rules

Ok, that makes perfect sense. Thanks, I'll definitely use both tags at once to return those results when needed for 2.1aa.

- The extraneous Selenium.Axe.AxeResult output line you're seeing is because of the Console.WriteLines of the result object; the result object doesn't currently provide a custom ToString, so these lines aren't going to be useful.

Ahh ok, that makes sense. Originally I was facing some difficulty when attempting to output the scan results to the console. I wanted to see all the errors. I wanted to see the results during regression testing to make sure no scan complaints appears and to be aware of any new a11y issues triggered before performing another full-site scan. But then informed me of using Fluent Assertions and that solved the issue. Thanks again on that. I'll remove the `Console.WriteLine(results);`

- I checked out the repo you linked, modified the test to scan for "wcag2a" issues rather than the "wcag21aa" it was originally using, and it looks like the test is finding the expected set of object-alt violations on the page under test (which is a level A rule)

Yeah, I was testing the "wcag2a" tag and I wasn't getting any scan results. I did, however, get results using "wcag2aa" tag only. Then I decided to test "wcag21AA" by itself, however, I was lazy and added the "wcag21aa" tag to the `AxeScanLevelA` method instead of the method I created for it. "wcag2aa" and "wcag21aa" do work well together for the desired output. I appreciate your over and beyond efforts and clarification.

- I'm not sure this has answered your question, but it looks like everything is working by design to me. Can you help me understand more concretely exactly what you believe the problem to be?

Yes, you did answer my question. 
Now when I run the code on my side I'm getting the output I need using the "wcag2a" tag. Odd that it wasn't working before since "wcag2aa" was, in fact, the only one working. Perhaps my connection or environment was off, maybe the Axe API was having some latency. Who knows.

I do truly appreciate your help here. 

Closing this issue as a user configuration issue.

Thanks @dbjorge
